app/models/katello/activation_key.rb

```diff
@@ -71,12 +75,84 @@ class ActivationKey < Katello::Model
+    end
+
+    def single_content_view_environment?
+      # also returns false if there are no content view environments
+      content_view_environments.size == 1
+    end
+
+    def single_content_view
+      if multi_content_view_environment?
+        Rails.logger.warn _("Activation key %s has more than one content view. Use #content_views instead.") % name
+      end
+      content_view_environments&.first&.content_view
+    end
+
+    def content_view
+      single_content_view
+    end
+
+    def environment
+      single_lifecycle_environment
+    end
+
+    def single_lifecycle_environment
+      if multi_content_view_environment?
+        Rails.logger.warn _("Activation key %s has more than one lifecycle environment. Use #lifecycle_environments instead.") % name
       end
+      content_view_environments&.first&.lifecycle_environment
+    end
+
+    # rubocop:disable Metrics/CyclomaticComplexity
+    def assign_single_environment(
+      content_view_id: nil, lifecycle_environment_id: nil, environment_id: nil,
+      content_view: nil, lifecycle_environment: nil, environment: nil
+    )
+      lifecycle_environment_id ||= environment_id || lifecycle_environment&.id || environment&.id || self.single_lifecycle_environment&.id
+      content_view_id ||= content_view&.id || self.single_content_view&.id
+
+      unless lifecycle_environment_id
+        fail _("Lifecycle environment must be specified")
+      end
+
+      unless content_view_id
+        fail _("Content view must be specified")
+      end
+
+      content_view_environment = ::Katello::ContentViewEnvironment
+        .where(:content_view_id => content_view_id, :environment_id => lifecycle_environment_id)
+        .first_or_create do |cve|
+        Rails.logger.info("ContentViewEnvironment not found for content view '#{cve.content_view_name}' and environment '#{cve.environment&.name}'; creating a new one.")
+      end
+      fail _("Unable to create ContentViewEnvironment. Check the logs for more information.") if content_view_environment.nil?
+
+      self.content_view_environments = [content_view_environment]
+    end
+
+    def self.in_environments(lifecycle_environments)
+      in_content_views_and_environments(:lifecycle_environments => lifecycle_environments)
+    end
+
+    def self.in_content_views_and_environments(content_views: nil, lifecycle_environments: nil, content_view_environments: nil)
+      relation = self.joins(:content_view_environment_activation_keys => :content_view_environment)
+      relation = relation.where("#{::Katello::ContentViewEnvironment.table_name}.content_view_id" => content_views) if content_views
+      relation = relation.where("#{::Katello::ContentViewEnvironment.table_name}.environment_id" => lifecycle_environments) if lifecycle_environments
+      relation = relation.where("#{::Katello::ContentViewEnvironment.table_name}.id" => content_view_environments) if content_view_environments
+      relation
```

Comment:
I've always had a hunch `in_content_views_and_environments` was a mistake. We're not really making clear here that you can only pass one of the three arguments, and there's an unintentional hierarchy where the last one wins (`content_view_environments` > `content_views` > `lifecycle_environments`).

What are your thoughts on refactoring this? I think single-purpose methods would make more sense:

`in_content_view_environments(content_view_environments:)` 
`in_content_views(content_views:)`
`in_lifecycle_environments(lifecycle_environments:)`

Then use each as needed.
I suppose this entire comment would also apply to content facets.

Reply:
update: I'm seeing now that since we use `relation.where` instead of `self.where`, it would be additive. Still confusing, though, and I still think we should refactor.